### PR TITLE
Update to newest version of `rust-mbedtls`

### PR DESCRIPTION
**Pull request by the-real-adammork**
Update to newest version of `rust-mbedtls`. Use heapless 0.8 to solve dependency issue where newer mbedtls needs spin 0.9.4.

### Motivation

Some assembly code from MbedTLS would break when compiled with "newer" versions of clang bundled with Xcode 13/14. The upstream fix was merged to `rust-mbedtls` this PR brings that change to `mobilecoin`

### Future Work
* Bring change to `libmobilecoin` 

[i still make hiphop lol - lvusm](https://soundcloud.com/lvusmparttwo/i-still-make-hiphop-lol)

### CODE REVIEW

**Review by eranrund — APPROVED** [review]
LGTM, based on https://github.com/mobilecoinfoundation/mobilecoin/pull/2962#issuecomment-1353608369, looks like once it gets merged you should go to https://github.com/mobilecoinofficial/android-bindings/blob/main/Cargo.toml (and likely full-service too?) to update the mbedtls version
Seems like at least with `android-bindings` it should be an easy uprev.

File: util/vec-map/Cargo.toml [hunk 1]
```diff
@@ -8,4 +8,4 @@ readme = "README.md"
 
 [dependencies]
 displaydoc = "0.2"
-heapless = { version = "0.8", default-features = false }
+heapless = { git = "https://github.com/japaric/heapless", branch = "main", default-features = false }
```
Comment (eranrund) [inline util/vec-map/Cargo.toml:11]: Oh, why is this needed?
At the very least we should point at a specific commit and not a branch.

**Review by eranrund — CHANGES_REQUESTED** [review]

File: util/vec-map/Cargo.toml (hunk 1 above)
Comment (the-real-adammork) [reply util/vec-map/Cargo.toml:11]: 0.8.0 isn't available in the crates.io, ok Ill update to a commit hash

Comment (eranrund) [reply util/vec-map/Cargo.toml]: What was the reason this needed to be updated in the first place?

Comment (jcape) [reply util/vec-map/Cargo.toml]: Couple other points:
1. It shouldn't track main, which presumably will change over time (i.e. it should track the diff that introduces this)
2. The recommended way to do this is to patch the crates.io registry to use a git commit at a specific revision via the workspace `Cargo.toml` files (e.g. at the top-level and `<thing>/enclave/trusted/Cargo.toml`)

Comment (the-real-adammork) [reply util/vec-map/Cargo.toml]: > hat was the reason this needed to be updated in the first place?

ya, Ive been trying to repro the "dependency issue" that made me update `heapless` but have been unsuccessful. Ill try a few more but maybe we don't need it.

Comment (the-real-adammork) [reply util/vec-map/Cargo.toml]: can't repro the dependency issue that I originally had that motivated these other changes, so I've removed them and force pushed just the mbedtls commit hash change

File: util/vec-map/Cargo.toml [hunk 2]
```diff
@@ -8,4 +8,9 @@ readme = "README.md"
 
 [dependencies]
 displaydoc = "0.2"
-heapless = { version = "0.7", default-features = false }
+heapless = { version = "^0.8", default-features = false }
+
+[patch.crates-io]
+# latest mbedtls needs spin `^0.9.4`, but `mc-util-vec-map` resolves spin to `^0.9.2` through `heapless` `^0.7`,
+# This specifies we use the latest version of heapless ~`0.9.4` to solve the dependency constraints. 
+heapless = { git = "https://github.com/mobilecoinofficial/heapless", rev = "2726f63bdc767d025f370d88341b1eb785178f2b", default-features = false }
```
Comment (jcape) [inline util/vec-map/Cargo.toml:16]: This patch doesn't do anything.
```suggestion
```

File: Cargo.toml [hunk 3]
```diff
@@ -241,3 +238,7 @@ x25519-dalek = { git = "https://github.com/mobilecoinfoundation/x25519-dalek.git
 
 # Override diesel dependency with our fork, to statically link SQLite.
 diesel = { git = "https://github.com/mobilecoinofficial/diesel", rev = "026f6379715d27c8be48396e5ca9059f4a263198" }
+
+# latest mbedtls needs spin `^0.9.4`, but `mc-util-vec-map` resolves spin to `^0.9.2` through `heapless` `^0.7`,
+# This specifies we use the latest version of heapless ~`0.9.4` to solve the dependency constraints. 
+heapless = { git = "https://github.com/mobilecoinofficial/heapless", rev = "2726f63bdc767d025f370d88341b1eb785178f2b", default-features = false }
```
Comment (jcape) [inline Cargo.toml:244]: Feature specification doesn't have any effect in a crates.io patch.

```suggestion
heapless = { git = "https://github.com/mobilecoinofficial/heapless", rev = "2726f63bdc767d025f370d88341b1eb785178f2b" }
```

**Review by jcape — APPROVED** [review]
LGTM with some nitpicks

File: util/vec-map/Cargo.toml (hunk 1 above)
Comment (the-real-adammork) [reply util/vec-map/Cargo.toml]: This is the dependency error if I only update mbedTLS.

```console
error: failed to select a version for `spin`.
    ... required by package `mbedtls v0.8.1 (https://github.com/mobilecoinfoundation/rust-mbedtls.git?rev=6f5bd68166a2c12fa3922867a6a6bd68904243d3#6f5bd681)`
    ... which satisfies dependency `mbedtls = "^0.8.1"` of package `mc-attest-net v4.0.0 (/__w/mobilecoin/mobilecoin/attest/net)`
    ... which satisfies path dependency `mc-attest-net` (locked to 4.0.0) of package `mc-attest-ake v4.0.0 (/__w/mobilecoin/mobilecoin/attest/ake)`
    ... which satisfies path dependency `mc-attest-ake` (locked to 4.0.0) of package `mc-attest-api v4.0.0 (/__w/mobilecoin/mobilecoin/attest/api)`
    ... which satisfies path dependency `mc-attest-api` (locked to 4.0.0) of package `mc-connection v4.0.0 (/__w/mobilecoin/mobilecoin/connection)`
    ... which satisfies path dependency `mc-connection` (locked to 4.0.0) of package `mc-connection-test-utils v4.0.0 (/__w/mobilecoin/mobilecoin/connection/test-utils)`
    ... which satisfies path dependency `mc-connection-test-utils` (locked to 4.0.0) of package `mc-ledger-sync v4.0.0 (/__w/mobilecoin/mobilecoin/ledger/sync)`
    ... which satisfies path dependency `mc-ledger-sync` (locked to 4.0.0) of package `mc-consensus-service v4.0.0 (/__w/mobilecoin/mobilecoin/consensus/service)`
versions that meet the requirements `^0.9.4` are: 0.9.4

all possible versions conflict with previously selected packages.

  previously selected package `spin v0.9.3`
    ... which satisfies dependency `spin = "^0.9.2"` (locked to 0.9.3) of package `heapless v0.7.16`
    ... which satisfies dependency `heapless = "^0.7"` (locked to 0.7.16) of package `mc-util-vec-map v4.0.0 (/__w/mobilecoin/mobilecoin/util/vec-map)`
    ... which satisfies path dependency `mc-util-vec-map` (locked to 4.0.0) of package `mc-transaction-extra v4.0.0 (/__w/mobilecoin/mobilecoin/transaction/extra)`
    ... which satisfies path dependency `mc-transaction-extra` (locked to 4.0.0) of package `mc-api v4.0.0 (/__w/mobilecoin/mobilecoin/api)`
    ... which satisfies path dependency `mc-api` (locked to 4.0.0) of package `mc-consensus-api v4.0.0 (/__w/mobilecoin/mobilecoin/consensus/api)`
    ... which satisfies path dependency `mc-consensus-api` (locked to 4.0.0) of package `mc-connection v4.0.0 (/__w/mobilecoin/mobilecoin/connection)`
    ... which satisfies path dependency `mc-connection` (locked to 4.0.0) of package `mc-connection-test-utils v4.0.0 (/__w/mobilecoin/mobilecoin/connection/test-utils)`
    ... which satisfies path dependency `mc-connection-test-utils` (locked to 4.0.0) of package `mc-ledger-sync v4.0.0 (/__w/mobilecoin/mobilecoin/ledger/sync)`
    ... which satisfies path dependency `mc-ledger-sync` (locked to 4.0.0) of package `mc-consensus-service v4.0.0 (/__w/mobilecoin/mobilecoin/consensus/service)`
```

the full fix now depends [on a fork of `heapless`](https://github.com/mobilecoinofficial/heapless/tree/adam/update-spin) where I increment the version of `spin` to `0.9.4`.

**Comment by the-real-adammork** [comment]
@jcape thank you for the suggestions, I accepted both. Also left a new comment on a older thread with the error output for @eranrund 's question.

**Review by eranrund — APPROVED** [review]
Suggest submitting a PR to https://github.com/japaric/heapless with the `spin` update, it appears actively maintained.

**Comment by the-real-adammork** [comment]
- [heapless PR](https://github.com/japaric/heapless/pull/337)
-
